ssl/tls13_enc.c

```diff
@@ -764,12 +772,132 @@ int tls13_change_cipher_state(SSL *s, int which)
         s->statem.enc_write_state = ENC_WRITE_STATE_WRITE_PLAIN_ALERTS;
     else
         s->statem.enc_write_state = ENC_WRITE_STATE_VALID;
+#ifndef OPENSSL_NO_KTLS
+# if defined(OPENSSL_KTLS_TLS13)
+#  ifndef __FreeBSD__
+    if (!(which & SSL3_CC_WRITE)
+        || ((which & SSL3_CC_WRITE) && (s->mode & SSL_MODE_NO_KTLS_TX)))
+        goto skip_ktls;
+
+    /* ktls supports only the maximum fragment size */
+    if (ssl_get_max_send_fragment(s) != SSL3_RT_MAX_PLAIN_LENGTH)
+        goto skip_ktls;
+
+    /* check whether cipher is known */
+    if (cipher == NULL)
+        goto skip_ktls;
+
+    /* check that cipher is AES_GCM_128, AES_GCM_256, AES_CCM_128 */
+    switch (EVP_CIPHER_nid(cipher))
+    {
+#   ifdef OPENSSL_KTLS_AES_GCM_128
+    case NID_aes_128_gcm:
+        if (EVP_CIPHER_mode(cipher) != EVP_CIPH_GCM_MODE ||
+            EVP_CIPHER_key_length(cipher) != TLS_CIPHER_AES_GCM_128_KEY_SIZE)
+          goto skip_ktls;
+        break;
+#   endif
+#   ifdef OPENSSL_KTLS_AES_GCM_256
+    case NID_aes_256_gcm:
+        if (EVP_CIPHER_mode(cipher) != EVP_CIPH_GCM_MODE ||
+            EVP_CIPHER_key_length(cipher) != TLS_CIPHER_AES_GCM_256_KEY_SIZE)
+          goto skip_ktls;
+        break;
+#   endif
+#   ifdef OPENSSL_KTLS_AES_CCM_128
+    case NID_aes_128_ccm:
+        if (EVP_CIPHER_mode(cipher) != EVP_CIPH_CCM_MODE ||
+            EVP_CIPHER_key_length(cipher) != TLS_CIPHER_AES_CCM_128_KEY_SIZE ||
+            EVP_CIPHER_CTX_tag_length(ciph_ctx) != EVP_CCM_TLS_TAG_LEN)
+          goto skip_ktls;
+        break;
+#   endif
+    default:
+        goto skip_ktls;
+    }
+
+    bio = s->wbio;
+
+    if (!ossl_assert(bio != NULL)) {
+        SSLfatal(s, SSL_AD_INTERNAL_ERROR, SSL_F_TLS13_CHANGE_CIPHER_STATE,
+                 ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+
+    /* All future data will get encrypted by ktls. Flush the BIO or skip ktls */
+    if (BIO_flush(bio) <= 0)
+        goto skip_ktls;
+
+    /* ktls doesn't support renegotiation */
+    if (BIO_get_ktls_send(bio)) {
```

Comment:
I'm trying to reconcile the comment, the code, and the fact that TLSv1.3 doesn't support renegotiation - and I'm failing.

Reply:
That check means that if `BIO_CTRL_GET_KTLS_SEND` flag is already set in BIO then `tls13_change_cipher_state` is called in case of renegotiation but not in initial handshake and that is not possible in Kernel TLS mode. The same check exists in TLSv1.2 code:
`    /* ktls doesn't support renegotiation */
    if ((BIO_get_ktls_send(s->wbio) && (which & SSL3_CC_WRITE)) ||
        (BIO_get_ktls_recv(s->rbio) && (which & SSL3_CC_READ))) {
        SSLfatal(s, SSL_AD_NO_RENEGOTIATION, SSL_F_TLS1_CHANGE_CIPHER_STATE,
                 ERR_R_INTERNAL_ERROR);
        goto err;
    }
`
It's a bit comples because of RX and TX path checks, in TLSv1.3 with support of only TX path the check is simplified.

Reply:
removed this check.